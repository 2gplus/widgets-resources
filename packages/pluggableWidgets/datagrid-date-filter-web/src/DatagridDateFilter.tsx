import { createElement, ReactElement } from "react";

import "./ui/DatagridDateFilter.scss";
import { FilterComponent } from "./components/FilterComponent";
import { useFilterDispatcher } from "./utils/provider";
import { DatagridDateFilterContainerProps } from "../typings/DatagridDateFilterProps";
import { registerLocale } from "react-datepicker";
import * as locales from "date-fns/locale";

declare type window = {
    mx: {
        session: {
            getConfig: () => {
                locale: {
                    languageTag: string;
                    patterns: {
                        date: string;
                    };
                };
            };
        };
    };
};

interface Locale {
    [key: string]: object;
}

declare let window: window;
const { languageTag = "en-US", patterns } = window.mx.session.getConfig().locale;

export default function DatagridDateFilter(props: DatagridDateFilterContainerProps): ReactElement {
    const filterDispatcher = useFilterDispatcher();

    const [language] = languageTag.split("-");
    const languageTagWithoutDash = languageTag.replace("-", "");

    if (languageTagWithoutDash in locales) {
        registerLocale(language, (locales as Locale)[languageTagWithoutDash]);
    } else if (language in locales) {
        registerLocale(language, (locales as Locale)[language]);
    }

    return (
        <FilterComponent
            adjustable={props.adjustable}
            defaultFilter={props.defaultFilter}
            dateFormat={patterns.date}
            filterDispatcher={filterDispatcher}
            locale={language}
            name={props.name}
            placeholder={props.placeholder?.value}
            screenReaderButtonCaption={props.screenReaderButtonCaption?.value}
            screenReaderInputCaption={props.screenReaderInputCaption?.value}
            tabIndex={props.tabIndex}
            defaultValue={props.defaultValue?.value}
        />
    );
}
