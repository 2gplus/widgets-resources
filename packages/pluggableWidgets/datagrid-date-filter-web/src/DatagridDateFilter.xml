<?xml version="1.0" encoding="utf-8"?>
<widget id="com.mendix.widget.web.datagriddatefilter.DatagridDateFilter" pluginWidget="true" offlineCapable="true"
        xmlns="http://www.mendix.com/widget/1.0/" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
        xsi:schemaLocation="http://www.mendix.com/widget/1.0/ ../../../../node_modules/mendix/custom_widget.xsd">
    <name>Date filter</name>
    <description/>
    <studioProCategory>Data controls</studioProCategory>
    <studioCategory>Data controls</studioCategory>
    <properties>
        <propertyGroup caption="General">
            <propertyGroup caption="General">
                <property key="defaultValue" type="expression" required="false">
                    <caption>Default value</caption>
                    <description />
                    <returnType type="DateTime" />
                </property>
                <property key="defaultFilter" type="enumeration" defaultValue="equal">
                    <caption>Default filter</caption>
                    <description />
                    <enumerationValues>
                        <enumerationValue key="greater">Greater than</enumerationValue>
                        <enumerationValue key="greaterEqual">Greater than or equal</enumerationValue>
                        <enumerationValue key="equal">Equal</enumerationValue>
                        <enumerationValue key="notEqual">Not equal</enumerationValue>
                        <enumerationValue key="smaller">Smaller than</enumerationValue>
                        <enumerationValue key="smallerEqual">Smaller than or equal</enumerationValue>
                    </enumerationValues>
                </property>
                <property key="placeholder" type="textTemplate" required="false">
                    <caption>Placeholder</caption>
                    <description />
                </property>
                <property key="adjustable" type="boolean" defaultValue="true">
                    <caption>Adjustable by user</caption>
                    <description />
                </property>
            </propertyGroup>
        </propertyGroup>
        <propertyGroup caption="Accessibility">
            <propertyGroup caption="Screen reader">
                <property key="screenReaderButtonCaption" type="textTemplate" required="false">
                    <caption>Button caption</caption>
                    <description />
                </property>
                <property key="screenReaderInputCaption" type="textTemplate" required="false">
                    <caption>Input caption</caption>
                    <description />
                </property>
            </propertyGroup>
        </propertyGroup>
    </properties>
</widget>
