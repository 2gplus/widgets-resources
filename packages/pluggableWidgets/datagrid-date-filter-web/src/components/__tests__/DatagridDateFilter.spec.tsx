import { Alert, FilterContextValue, ListAttributeValueBuilder } from "@mendix/piw-utils-internal";
import { mount } from "enzyme";
import { createContext, createElement } from "react";
import DatagridDateFilter from "../../DatagridDateFilter";

const commonProps = {
    class: "filter-test",
    tabIndex: 0,
    name: "filter-test",
    defaultFilter: "equal" as const,
    adjustable: true
};

const mxObject = {
    session: {
        getConfig: () => ({
            locale: {
                languageTag: "en-US",
                patterns: {
                    date: "dd/MM/YYYY"
                }
            }
        })
    }
};

describe("Date Filter", () => {
    describe("with single attribute", () => {
        beforeAll(() => {
            (window as any)["com.mendix.widgets.web.filterable.filterContext"] = createContext({
                filterDispatcher: jest.fn(),
                singleAttribute: new ListAttributeValueBuilder().withType("DateTime").withFilterable(true).build()
            } as FilterContextValue);
            (window as any).mx = mxObject;

            jest.spyOn(global.Math, "random").mockReturnValue(0.123456789);
        });

        it("renders correctly", () => {
            const filter = mount(<DatagridDateFilter {...commonProps} />);

            expect(filter).toMatchSnapshot();
        });

        afterAll(() => {
            (window as any)["com.mendix.widgets.web.filterable.filterContext"] = undefined;
        });
    });

    describe("with multiple attributes", () => {
        beforeAll(() => {
            (window as any)["com.mendix.widgets.web.filterable.filterContext"] = createContext({
                filterDispatcher: jest.fn(),
                multipleAttributes: {
                    attribute1: new ListAttributeValueBuilder()
                        .withId("attribute1")
                        .withType("DateTime")
                        .withFilterable(true)
                        .build(),
                    attribute2: new ListAttributeValueBuilder()
                        .withId("attribute2")
                        .withType("DateTime")
                        .withFilterable(true)
                        .build()
                }
            } as FilterContextValue);
            (window as any).mx = mxObject;

            jest.spyOn(global.Math, "random").mockReturnValue(0.123456789);
        });

        it("renders correctly", () => {
            const filter = mount(<DatagridDateFilter {...commonProps} />);

            expect(filter).toMatchSnapshot();
        });

        afterAll(() => {
            (window as any)["com.mendix.widgets.web.filterable.filterContext"] = undefined;
        });
    });

    describe("with wrong attribute's type", () => {
        beforeAll(() => {
            (window as any)["com.mendix.widgets.web.filterable.filterContext"] = createContext({
                filterDispatcher: jest.fn(),
                singleAttribute: new ListAttributeValueBuilder().withType("Decimal").withFilterable(true).build()
            } as FilterContextValue);
            (window as any).mx = mxObject;
        });

        it("renders error message", () => {
            const filter = mount(<DatagridDateFilter {...commonProps} />);

            expect(filter.find(Alert).text()).toBe(
                "The attribute type being used for Date filter is not 'Date and time'"
            );
        });

        afterAll(() => {
            (window as any)["com.mendix.widgets.web.filterable.filterContext"] = undefined;
        });
    });

    describe("with wrong multiple attributes' types", () => {
        beforeAll(() => {
            (window as any)["com.mendix.widgets.web.filterable.filterContext"] = createContext({
                filterDispatcher: jest.fn(),
                multipleAttributes: {
                    attribute1: new ListAttributeValueBuilder()
                        .withId("attribute1")
                        .withType("String")
                        .withFilterable(true)
                        .build(),
                    attribute2: new ListAttributeValueBuilder()
                        .withId("attribute2")
                        .withType("Decimal")
                        .withFilterable(true)
                        .build()
                }
            } as FilterContextValue);
            (window as any).mx = mxObject;
        });

        it("renders error message", () => {
            const filter = mount(<DatagridDateFilter {...commonProps} />);

            expect(filter.find(Alert).text()).toBe(
                'To use multiple filters you need to define a filter identification in the properties of Date filter or have a "Date and Time" attribute available.'
            );
        });

        afterAll(() => {
            (window as any)["com.mendix.widgets.web.filterable.filterContext"] = undefined;
        });
    });

    describe("with no context", () => {
        beforeAll(() => {
            (window as any)["com.mendix.widgets.web.filterable.filterContext"] = undefined;
            (window as any).mx = mxObject;
        });

        it("renders error message", () => {
            const filter = mount(<DatagridDateFilter {...commonProps} />);

            expect(filter.find(Alert).text()).toBe(
                "The Date filter widget must be placed inside the header of the Data grid 2.0 or Gallery widget."
            );
        });
    });
});
