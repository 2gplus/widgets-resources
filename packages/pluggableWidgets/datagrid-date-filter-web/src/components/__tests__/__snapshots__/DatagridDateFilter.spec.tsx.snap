// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Date Filter with multiple attributes renders correctly 1`] = `
<DatagridDateFilter
  adjustable={true}
  class="filter-test"
  defaultFilter="equal"
  name="filter-test"
  tabIndex={0}
>
  <FilterComponent
    adjustable={true}
    className="filter-test"
    dateFormat="dd/MM/YYYY"
    defaultFilter="equal"
    locale="en"
    name="filter-test"
    tabIndex={0}
    updateFilters={[Function]}
  >
    <div
      className="filter-container filter-test"
      data-focusindex={0}
    >
      <FilterSelector
        defaultFilter="equal"
        name="filter-test"
        onChange={[Function]}
        options={
          Array [
            Object {
              "label": "Greater than",
              "value": "greater",
            },
            Object {
              "label": "Greater than or equal",
              "value": "greaterEqual",
            },
            Object {
              "label": "Equal",
              "value": "equal",
            },
            Object {
              "label": "Not equal",
              "value": "notEqual",
            },
            Object {
              "label": "Smaller than",
              "value": "smaller",
            },
            Object {
              "label": "Smaller than or equal",
              "value": "smallerEqual",
            },
          ]
        }
      >
        <div
          className="filter-selector"
        >
          <div
            className="filter-selector-content"
          >
            <button
              aria-controls="filter-test-filter-selectors"
              aria-expanded={false}
              aria-haspopup={true}
              className="btn btn-default filter-selector-button button-icon equal"
              onClick={[Function]}
              onKeyDown={[Function]}
            >
               
            </button>
          </div>
        </div>
      </FilterSelector>
      <ForwardRef
        adjustable={true}
        dateFormat="dd/MM/YYYY"
        locale="en"
        name="filter-test"
        setValue={[Function]}
        value={null}
      >
        <Portal
          containerInfo={
            <body>
              <div
                id="datepicker_0.123456789"
                style="position: fixed;"
              />
              <div
                id="datepicker_0.123456789"
                style="position: fixed;"
              />
            </body>
          }
        >
          <div
            id="datepicker_0.123456789"
            style={
              Object {
                "position": "fixed",
              }
            }
          />
        </Portal>
        <span
          className="sr-only"
          id="filter-test-label"
        />
        <r
          allowSameDay={false}
          ariaLabelledBy="filter-test-label"
          autoFocus={false}
          className="form-control filter-input"
          customTimeInput={null}
          dateFormat="dd/MM/YYYY"
          dateFormatCalendar="LLLL yyyy"
          disabled={false}
          disabledKeyboardNavigation={false}
          dropdownMode="select"
          enableTabLoop={true}
          excludeScrollbar={true}
          focusSelectedMonth={false}
          locale="en"
          monthsShown={1}
          nextMonthButtonLabel="Next Month"
          nextYearButtonLabel="Next Year"
          onBlur={[Function]}
          onCalendarClose={[Function]}
          onCalendarOpen={[Function]}
          onChange={[Function]}
          onClickOutside={[Function]}
          onFocus={[Function]}
          onInputClick={[Function]}
          onInputError={[Function]}
          onKeyDown={[Function]}
          onMonthChange={[Function]}
          onSelect={[Function]}
          onYearChange={[Function]}
          open={false}
          popperModifiers={
            Array [
              Object {
                "name": "offset",
                "options": Object {
                  "offset": Array [
                    0,
                    0,
                  ],
                },
              },
            ]
          }
          popperPlacement="bottom-end"
          portalId="datepicker_0.123456789"
          preventOpenOnFocus={true}
          previousMonthButtonLabel="Previous Month"
          previousYearButtonLabel="Previous Year"
          readOnly={false}
          renderDayContents={[Function]}
          selected={null}
          shouldCloseOnSelect={false}
          showFourColumnMonthYearPicker={false}
          showFullMonthYearPicker={false}
          showMonthDropdown={true}
          showMonthYearPicker={false}
          showPopperArrow={false}
          showPreviousMonths={false}
          showQuarterYearPicker={false}
          showTimeInput={false}
          showTimeSelect={false}
          showTwoColumnMonthYearPicker={false}
          showYearDropdown={true}
          showYearPicker={false}
          strictParsing={true}
          timeCaption="Time"
          timeInputLabel="Time"
          timeIntervals={30}
          useWeekdaysShort={true}
          withPortal={false}
          yearItemNumber={12}
        >
          <r
            enableTabLoop={true}
            hidePopper={true}
            popperComponent={null}
            popperModifiers={
              Array [
                Object {
                  "name": "offset",
                  "options": Object {
                    "offset": Array [
                      0,
                      0,
                    ],
                  },
                },
              ]
            }
            popperOnKeyDown={[Function]}
            popperPlacement="bottom-end"
            popperProps={Object {}}
            portalId="datepicker_0.123456789"
            targetComponent={
              <div
                className="react-datepicker__input-container"
              >
                <input
                  aria-labelledby="filter-test-label"
                  autoFocus={false}
                  className="form-control filter-input"
                  disabled={false}
                  onBlur={[Function]}
                  onChange={[Function]}
                  onClick={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  readOnly={false}
                  type="text"
                  value=""
                />
              </div>
            }
          >
            <Manager
              className="react-datepicker-manager"
            >
              <Reference>
                <div
                  className="react-datepicker-wrapper"
                >
                  <div
                    className="react-datepicker__input-container"
                  >
                    <input
                      aria-labelledby="filter-test-label"
                      autoFocus={false}
                      className="form-control filter-input"
                      disabled={false}
                      onBlur={[Function]}
                      onChange={[Function]}
                      onClick={[Function]}
                      onFocus={[Function]}
                      onKeyDown={[Function]}
                      readOnly={false}
                      type="text"
                      value=""
                    />
                  </div>
                </div>
              </Reference>
            </Manager>
          </r>
        </r>
        <button
          aria-controls="datepicker_0.123456789"
          aria-expanded={false}
          aria-haspopup={true}
          aria-label="Show calendar"
          className="btn btn-default btn-calendar"
          onClick={[Function]}
          onKeyDown={[Function]}
        >
          <span
            className="glyphicon glyphicon-calendar"
          />
        </button>
      </ForwardRef>
    </div>
  </FilterComponent>
</DatagridDateFilter>
`;

exports[`Date Filter with single attribute renders correctly 1`] = `
<DatagridDateFilter
  adjustable={true}
  class="filter-test"
  defaultFilter="equal"
  name="filter-test"
  tabIndex={0}
>
  <FilterComponent
    adjustable={true}
    className="filter-test"
    dateFormat="dd/MM/YYYY"
    defaultFilter="equal"
    locale="en"
    name="filter-test"
    tabIndex={0}
    updateFilters={[Function]}
  >
    <div
      className="filter-container filter-test"
      data-focusindex={0}
    >
      <FilterSelector
        defaultFilter="equal"
        name="filter-test"
        onChange={[Function]}
        options={
          Array [
            Object {
              "label": "Greater than",
              "value": "greater",
            },
            Object {
              "label": "Greater than or equal",
              "value": "greaterEqual",
            },
            Object {
              "label": "Equal",
              "value": "equal",
            },
            Object {
              "label": "Not equal",
              "value": "notEqual",
            },
            Object {
              "label": "Smaller than",
              "value": "smaller",
            },
            Object {
              "label": "Smaller than or equal",
              "value": "smallerEqual",
            },
          ]
        }
      >
        <div
          className="filter-selector"
        >
          <div
            className="filter-selector-content"
          >
            <button
              aria-controls="filter-test-filter-selectors"
              aria-expanded={false}
              aria-haspopup={true}
              className="btn btn-default filter-selector-button button-icon equal"
              onClick={[Function]}
              onKeyDown={[Function]}
            >
               
            </button>
          </div>
        </div>
      </FilterSelector>
      <ForwardRef
        adjustable={true}
        dateFormat="dd/MM/YYYY"
        locale="en"
        name="filter-test"
        setValue={[Function]}
        value={null}
      >
        <Portal
          containerInfo={
            <body>
              <div
                id="datepicker_0.123456789"
                style="position: fixed;"
              />
            </body>
          }
        >
          <div
            id="datepicker_0.123456789"
            style={
              Object {
                "position": "fixed",
              }
            }
          />
        </Portal>
        <span
          className="sr-only"
          id="filter-test-label"
        />
        <r
          allowSameDay={false}
          ariaLabelledBy="filter-test-label"
          autoFocus={false}
          className="form-control filter-input"
          customTimeInput={null}
          dateFormat="dd/MM/YYYY"
          dateFormatCalendar="LLLL yyyy"
          disabled={false}
          disabledKeyboardNavigation={false}
          dropdownMode="select"
          enableTabLoop={true}
          excludeScrollbar={true}
          focusSelectedMonth={false}
          locale="en"
          monthsShown={1}
          nextMonthButtonLabel="Next Month"
          nextYearButtonLabel="Next Year"
          onBlur={[Function]}
          onCalendarClose={[Function]}
          onCalendarOpen={[Function]}
          onChange={[Function]}
          onClickOutside={[Function]}
          onFocus={[Function]}
          onInputClick={[Function]}
          onInputError={[Function]}
          onKeyDown={[Function]}
          onMonthChange={[Function]}
          onSelect={[Function]}
          onYearChange={[Function]}
          open={false}
          popperModifiers={
            Array [
              Object {
                "name": "offset",
                "options": Object {
                  "offset": Array [
                    0,
                    0,
                  ],
                },
              },
            ]
          }
          popperPlacement="bottom-end"
          portalId="datepicker_0.123456789"
          preventOpenOnFocus={true}
          previousMonthButtonLabel="Previous Month"
          previousYearButtonLabel="Previous Year"
          readOnly={false}
          renderDayContents={[Function]}
          selected={null}
          shouldCloseOnSelect={false}
          showFourColumnMonthYearPicker={false}
          showFullMonthYearPicker={false}
          showMonthDropdown={true}
          showMonthYearPicker={false}
          showPopperArrow={false}
          showPreviousMonths={false}
          showQuarterYearPicker={false}
          showTimeInput={false}
          showTimeSelect={false}
          showTwoColumnMonthYearPicker={false}
          showYearDropdown={true}
          showYearPicker={false}
          strictParsing={true}
          timeCaption="Time"
          timeInputLabel="Time"
          timeIntervals={30}
          useWeekdaysShort={true}
          withPortal={false}
          yearItemNumber={12}
        >
          <r
            enableTabLoop={true}
            hidePopper={true}
            popperComponent={null}
            popperModifiers={
              Array [
                Object {
                  "name": "offset",
                  "options": Object {
                    "offset": Array [
                      0,
                      0,
                    ],
                  },
                },
              ]
            }
            popperOnKeyDown={[Function]}
            popperPlacement="bottom-end"
            popperProps={Object {}}
            portalId="datepicker_0.123456789"
            targetComponent={
              <div
                className="react-datepicker__input-container"
              >
                <input
                  aria-labelledby="filter-test-label"
                  autoFocus={false}
                  className="form-control filter-input"
                  disabled={false}
                  onBlur={[Function]}
                  onChange={[Function]}
                  onClick={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  readOnly={false}
                  type="text"
                  value=""
                />
              </div>
            }
          >
            <Manager
              className="react-datepicker-manager"
            >
              <Reference>
                <div
                  className="react-datepicker-wrapper"
                >
                  <div
                    className="react-datepicker__input-container"
                  >
                    <input
                      aria-labelledby="filter-test-label"
                      autoFocus={false}
                      className="form-control filter-input"
                      disabled={false}
                      onBlur={[Function]}
                      onChange={[Function]}
                      onClick={[Function]}
                      onFocus={[Function]}
                      onKeyDown={[Function]}
                      readOnly={false}
                      type="text"
                      value=""
                    />
                  </div>
                </div>
              </Reference>
            </Manager>
          </r>
        </r>
        <button
          aria-controls="datepicker_0.123456789"
          aria-expanded={false}
          aria-haspopup={true}
          aria-label="Show calendar"
          className="btn btn-default btn-calendar"
          onClick={[Function]}
          onKeyDown={[Function]}
        >
          <span
            className="glyphicon glyphicon-calendar"
          />
        </button>
      </ForwardRef>
    </div>
  </FilterComponent>
</DatagridDateFilter>
`;
